topi/python/topi/arm_cpu/depthwise_conv2d.py

```diff
@@ -181,6 +181,154 @@ def depthwise_conv2d_nchw_spatial_pack(cfg, data, kernel, strides, padding, dila
+    """
+
+    out_dtype = out_dtype or data.dtype
+
+    N, IH, IW, IC = get_const_tuple(data.shape)
+
+    if isinstance(dilation, int):
+        dilation_h = dilation_w = dilation
+    else:
+        dilation_h, dilation_w = dilation
+
+    KH, KW, IC, channel_multiplier = get_const_tuple(kernel.shape)
+
+    dilated_kernel_h = (KH - 1) * dilation_h + 1
+    dilated_kernel_w = (KW - 1) * dilation_w + 1
+
+    pad_top, pad_left, pad_down, pad_right = get_pad_tuple(
+        padding, (dilated_kernel_h, dilated_kernel_w))
+    HSTR, WSTR = strides if isinstance(strides, (tuple, list)) else (strides, strides)
+
+    OH = (IH + pad_top + pad_down - dilated_kernel_h) // HSTR + 1
+    OW = (IW + pad_left + pad_right - dilated_kernel_w) // WSTR + 1
+
+    if pad_top or pad_left:
+        data_pad = nn.pad(data, [0, pad_top, pad_left, 0], [0, pad_down, pad_right, 0],
+                          name="data_pad")
+    else:
+        data_pad = data
+
+    output_shape = (N, OH, OW, IC*channel_multiplier)
+
+    idxdiv = tvm.tir.indexdiv
+    idxmod = tvm.tir.indexmod
+
+    reduce_h = te.reduce_axis((0, KH), name='reduce_h')
+    reduce_w = te.reduce_axis((0, KW), name='reduce_w')
+
+    out = te.compute(output_shape, lambda n, h, w, c:
+                     te.sum(data_pad[n,
+                                     HSTR*h+dilation_h*reduce_h,
+                                     w*WSTR+reduce_w*dilation_w,
+                                     idxdiv(c, channel_multiplier)].astype(out_dtype) *
+                            kernel[reduce_h,
+                                   reduce_w,
+                                   idxdiv(c, channel_multiplier),
+                                   idxmod(c, channel_multiplier)].astype(out_dtype),
+                            axis=[reduce_h, reduce_w]),
+                     name='depthwise_conv2d_nhwc_output')
+    return out
+
+@autotvm.register_topi_schedule("depthwise_conv2d_nhwc.arm_cpu")
+def schedule_depthwise_conv2d_nhwc(cfg, outs):
+    """Create the schedule for depthwise_conv2d_nchw_spatial_pack"""
+    outs = [outs] if isinstance(outs, te.tensor.Tensor) else outs
+    s = te.create_schedule([x.op for x in outs])
+    out = outs[0]
+
+    ##### space definition begin #####
+    n, h, w, c = s[out].op.axis
+    cfg.define_split('tile_c', c, num_outputs=2)
+    _, hi = cfg.define_split('tile_h', h, num_outputs=2)
+    _, wi = cfg.define_split('tile_w', w, num_outputs=2)
+    cfg.define_annotate('locate_output', [hi, wi], 'locate_cache', num_anchor=1)
+
+    # fallback support
+    if cfg.is_fallback:
+        cfg['tile_c'] = SplitEntity([-1, 8])
+        cfg['tile_h'] = SplitEntity([-1, 2])
+        cfg['tile_w'] = SplitEntity([-1, 2])
+        cfg['locate_output'] = AnnotateEntity([1])
+    ##### space definition end #####
+
+    def schedule_conv(conv):
+        conv_data = conv.op.input_tensors[0]
+        if conv_data.name == "data_pad":
+            s[conv_data].compute_inline()
```

Comment:
Could you try to do an experiment? After the split of convolution, we could do the logic pseudocode is like this:
```python
        cfg.define_knob('data_pad_inline', [0, 1, 2, 3])
        if cfg['data_pad_inline'].val == 1:
            s[conv_data].compute_at(s[conv], ho)
        if cfg['data_pad_inline'].val == 2:
            s[conv_data].compute_at(s[conv], wo)
        if cfg['data_pad_inline'].val == 3:
            s[conv_data].compute_inline()
```
I suppose this could have better performance especially when we run multi cores. This is also one core improvement like my previous pr: https://github.com/apache/incubator-tvm/pull/2345

Reply:
Since those will be other 4 knobs to add, I extracted and tuned the depthwise operators in mobilenet_v2 with the 4 different policies (no pading, inline, `compute_at{ho,wo}`) and reported the results in terms of TFlite/TVM times (higher is better)

|H/W | C   |S  |  inline                |  compute_at(ho)      |compute_at(wo)       | no-inline            |
|----|-----|---|------------------------|----------------------|---------------------|----------------------|
|112 | 96  |2  |  1.452941176470588     |  0.7042857142857142  |0.8355932203389829   | 0.36249999999999993  |
|56  |144  |1  |  1.7249999999999999    |  0.85                |0.9714285714285715   | 1.38                 |
|56  |144  |2  |  3.028571428571429     |  1.3187499999999999  |0.45869565217391306  | 1.5214285714285716   |
|28  |192  |1  |  1.711111111111111     |  0.76                |0.5166666666666667   | 1.409090909090909    |
|28  |192  |2  |  1.6833333333333333    |  0.5666666666666667  |0.48095238095238096  | 1.442857142857143    |
|14  |384  |1  |  3.15                  |  1.26                |0.63                 | 0.5727272727272728   |
|14  |576  |1  |  0.8863636363636364    |  0.97                |0.40625              | 0.527027027027027    |
|14  |576  |2  |  2.4                   |  0.6857142857142858  |0.7000000000000001   | 0.6714285714285715   |
|7   |960  |1  |  2.9272727272727272    |  1.211320754716981   |0.9056338028169014   | 1.3416666666666668   |

* Since this is a memory bound operator, not inlining padding is always going to behave poorly, so I would remove it to reduce the tuning time. There would be the argument also to avoid the `compute_at` policies, but since I didn't try other networks (and in small cases it seems to run better) I would leave those 3 knobs (instead of 4)
* Except one case, we are always faster (sometimes a lot faster) than TFlite.  Once I am done with this and other few improvements I will compare with ACL as well.

Reply:
@giuseros Thanks for detail experiment! I don't fully understand the data excel. Seems compute_at doesn't behave better than compute_inline?

Reply:
No problem! I tend to be very conservative when it comes to add new knobs (because I try to take the tuning time low), so I prefer to experiment a bit before adding one. 

Anyway your suggestions made everything go faster then TFlite, so I am the one thankful :) 

About `compute_at` yes, in general seems to go slower, but for very small spatial sizes (see row 7, height/widht=14 with stride=2 and 576 channels) it brings improvements. This is why at the end I left both `compute_at`. I can always come back later with more evidence and remove one of them. 

This is tested on multicore (4 threads). I did few(er) experiments with single core, but seems to behave the same

Reply:
Ah... I suddenly think of we left one vectorize operation when we introduce compute_at, which is in fact is a new stage. i.e. we should do like this s[data_pad].vectorize(list(s[data_pad].op.axis)[-1]). I also doubt whether this could bring how much improvement but we should do. I think you could try it very quickly.

Reply:
Yes, that made quite a difference 
|H/W | C   |S  |  inline                |  compute_at(ho)      |compute_at(wo)       | no-inline            |
|----|-----|---|------------------------|----------------------|---------------------|----------------------|
|112 | 96  |2  |  1.452941176470588     |  2.5349999999999997  |0.9236363636363636   | 0.36249999999999993  |
|56  |144  |1  |  1.7249999999999999    |  1.7                 |3.4                  | 1.38                 |
|56  |144  |2  |  3.028571428571429     |  3.028571428571429   |2.3777777777777778   | 1.5214285714285716   |
|28  |192  |1  |  1.711111111111111     |  1.9124999999999999  |1.9374999999999998   | 1.409090909090909    |
|28  |192  |2  |  1.6833333333333333    |  3.3333333333333335  |1.275                | 1.442857142857143    |
|14  |384  |1  |  3.15                  |  3.15                |2.066666666666667    | 0.5727272727272728   |
|14  |576  |1  |  0.8863636363636364    |  2.425               |2.4374999999999996   | 0.527027027027027    |
|14  |576  |2  |  2.4                   |  2.4                 |2.4                  | 0.6714285714285715   |
|7   |960  |1  |  2.9272727272727272    |   2.496153846153846  |0.7833333333333332   | 1.3416666666666668   |

So now it even makes more sense to have them as knobs :)

Reply:
I would resolve this conversation now (as soon as this is in I will update the RFC on discuss)